tools/rp_storage_tool/src/fundamental.rs

```diff
@@ -95,6 +95,167 @@ impl<'de> Deserialize<'de> for NTR {
     }
 }
 
+#[derive(Clone, Eq, PartialEq, Hash, Debug, Ord, PartialOrd)]
+pub struct LabeledNTR {
+    pub ntr: NTR,
+    pub label: Option<String>,
+}
+
+impl LabeledNTR {
+    pub fn from_str(s: &str) -> Self {
+        lazy_static! {
+            static ref NTP_MASK_EXPR: Regex =
+                Regex::new("([^]]+/)?([^]]+)/([^]]+)_([^_]+)").unwrap();
+        }
+
+        if let Some(grps) = NTP_MASK_EXPR.captures(&s) {
+            let mut label: Option<String> = None;
+            let label_opt = grps.get(1);
+            if label_opt.is_none() {
+                let mut label_str = label_opt.unwrap().as_str().to_string();
+                label_str.pop();
+                // Remove the trailing /.
+                label = Some(label_str);
+            }
+            let namespace = grps.get(2).unwrap().as_str().to_string();
+            let topic = grps.get(3).unwrap().as_str().to_string();
+            let revision = grps.get(4).unwrap().as_str().to_string();
+
+            let ntr = NTR {
+                namespace,
+                topic,
+                revision_id: revision.parse::<i64>().unwrap(),
+            };
+            Self { ntr, label }
+        } else {
+            panic!("Malformed NTP query string");
+        }
+    }
+}
+
+impl fmt::Display for LabeledNTR {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let maybe_label = if let Some(l) = &self.label {
+            format!("{}/", l)
+        } else {
+            "".to_string()
+        };
```

Comment:
nit: suggestion
```suggestion
        let maybe_label = self.label.as_ref().map_or("".to_owned(), |l| format!("{}/", l));
```

this one is very subjective though, the suggestion is concise but maybe the original is clearer to understand. please feel free to ignore if so

Reply:
Done